Sources/SafeDICore/Generators/ScopeGenerator.swift

```diff
@@ -236,42 +262,90 @@ actor ScopeGenerator {
+                for otherScope in propertiesToGenerate {
+                    if
+                        let property = otherScope.property,
+                        scope.requiredReceivedProperties.contains(property),
+                        let dependenciesCount = scopeToOutgoingUnresolvedDependenciesCount[scope]
+                    {
+                        scopeToOutgoingUnresolvedDependenciesCount[scope] = dependenciesCount + 1
+                        scopeToIncomingDependencyScopes[otherScope, default: []] += [scope]
+                    }
+                }
             }
-            if let lastDependencyToGenerate = propertyToGenerate
-                // Ignore our own property.
-                .dropFirst()
-                // Reverse the list to find the last property we depend on first.
-                .reversed()
-                .first(where: { futurePropertyToGenerate in
-                    if let futureProperty = futurePropertyToGenerate.property,
-                       propertyToGenerate
-                        .requiredReceivedProperties
-                        .contains(futureProperty)
-                    { true } else { false }
+
+            // Step 2: Topological Sort
+            var nextPropertiesToGenerate = scopeToOutgoingUnresolvedDependenciesCount
+                .filter { $0.value == 0 }
+                .map(\.key)
+                .sorted(by: {
+                    // Use original dependency order as our sorting key.
+                    scopeOriginalScopeIndex[$0]! < scopeOriginalScopeIndex[$1]!
                 })
-            {
-                // We depend on (at least) one item further ahead in the list!
-                // Make sure we are created after our dependencies.
-                lastDependencyToGenerate.insert(propertyToGenerate.value)
-                if let head = propertyToGenerate.remove() {
-                    orderedPropertiesToGenerate = head
+            var orderedPropertiesToGenerate = [ScopeGenerator]()
+            while !nextPropertiesToGenerate.isEmpty {
+                let nextProperty = nextPropertiesToGenerate.removeFirst()
+                orderedPropertiesToGenerate.append(nextProperty)
+
+                if let incomingDependencyScopes = scopeToIncomingDependencyScopes[nextProperty] {
+                    for dependantScope in incomingDependencyScopes {
+                        if let dependenciesCount = scopeToOutgoingUnresolvedDependenciesCount[dependantScope] {
+                            let newDependenciesCount = dependenciesCount - 1
+                            scopeToOutgoingUnresolvedDependenciesCount[dependantScope] = newDependenciesCount
+
+                            if newDependenciesCount == 0 {
+                                nextPropertiesToGenerate.append(dependantScope)
+                            }
+                        }
+                    }
+                }
+            }
+
+            // Step 3: Check for cycle
```

Comment:
we were detecting cycles in instantiated properties but not received properties!!